### ISSUE TITLE
Segmentation Fault for extract_voxel_grid

### ISSUE THREAD

**Issue by hooyahliu** [isl-org/Open3D#6712]
### Checklist

- [x] I have searched for [similar issues](https://github.com/isl-org/Open3D/issues).
- [X] For Python issues, I have tested with the [latest development wheel](https://www.open3d.org/docs/latest/getting_started.html#development-version-pip).
- [X] I have checked the [release documentation](https://www.open3d.org/docs/release/) and the [latest documentation](https://www.open3d.org/docs/latest/) (for `main` branch).

### Describe the issue

When using Open3D 0.18.0 to execute self._volume.extract_voxel_grid() on the generated TSDF volume, the program exits with the error Segmentation fault (core dumped).
But using 0.12.0 no error occurs.

### Steps to reproduce the bug

```python
class TSDFVolume(object):
    def __init__(self, size, resolution):
        self.size = size
        self.resolution = resolution
        self.voxel_size = self.size / self.resolution
        self.sdf_trunc = 4 * self.voxel_size

        self._volume = o3d.pipelines.integration.UniformTSDFVolume(
            length=self.size,
            resolution=self.resolution,
            sdf_trunc=self.sdf_trunc,
            color_type=o3d.pipelines.integration.TSDFVolumeColorType.NoColor,
        )

    def integrate(self, depth_img, intrinsic, extrinsic):
        """
        Args:
            depth_img: The depth image.
            intrinsic: The intrinsic parameters of a pinhole camera model.
            extrinsics: The transform from the TSDF to camera coordinates, T_eye_task.
        """
        rgbd = o3d.geometry.RGBDImage.create_from_color_and_depth(
            o3d.geometry.Image(np.empty_like(depth_img)),
            o3d.geometry.Image(depth_img),
            depth_scale=1.0,
            depth_trunc=2.0,
            convert_rgb_to_intensity=False,
        )

        intrinsic = o3d.camera.PinholeCameraIntrinsic(
            width=intrinsic.width,
            height=intrinsic.height,
            fx=intrinsic.fx,
            fy=intrinsic.fy,
            cx=intrinsic.cx,
            cy=intrinsic.cy,
        )

        extrinsic = extrinsic.as_matrix()
        self._volume.integrate(rgbd, intrinsic, extrinsic)

    def get_grid(self):
        shape = (1, self.resolution, self.resolution, self.resolution)
        tsdf_grid = np.zeros(shape, dtype=np.float32)
        voxels = self._volume.extract_voxel_grid().get_voxels()
        for voxel in voxels:
            i, j, k = voxel.grid_index
            tsdf_grid[0, i, j, k] = voxel.color[0]
        return tsdf_grid
```

### Error message

Segmentation fault (core dumped)

### Expected behavior

Return voxels.

### Open3D, Python and System information

```markdown
- Operating system: Ubuntu 20.04 
- Python version: Python 3.8 
- Open3D version: 0.18.0 0.12.0
- System architecture: x86
- Is this a remote workstation?: no
- How did you install Open3D?: pip
```

**Comment by nicolaloi** [comment]
When running the tutorial `/examples/python/pipelines/rgbd_integration_uniform.py` I get a segmentation fault or the program hangs indefinitely at `voxel_grid = volume.extract_voxel_grid()`.

The probable cause is that `VoxelGrid` is a `std::unordered_map`, which is not thread-safe, so OpenMP hangs forever or throws a segfault. Problematic line: `voxel_grid->voxels_[index] = geometry::Voxel(index, color);` in `cpp/open3d/pipelines/integration/UniformTSDFVolume.cpp`.

An easy fix would be to remove the OpenMP parallelization (fix 1.). Other possible fixes are (fix 2.) use `tbb:concurrent_unordered_map` which is thread-safe, but then you have to convert it back to a `std::unordered_map` (which I presume is O(n)?), or (fix 3.) use multiple `std::unordered_map`s, one for each thread, and then merge them after the parallelization (which would be O(n) average case). I have tested fix 1. and fix 2. for multiple voxel resolutions (16, 32, 64, 128, 256, 512, 768), and their timings are practically the same. @ssheorey do you have any other ideas?

**Comment by hooyahliu** [comment]
> When running the tutorial `/examples/python/pipelines/rgbd_integration_uniform.py` I get a segmentation fault or the program hangs indefinitely at `voxel_grid = volume.extract_voxel_grid()`.
> 
> The probable cause is that `VoxelGrid` is a `std::unordered_map`, which is not thread-safe, so OpenMP hangs forever or throws a segfault. Problematic line: `voxel_grid->voxels_[index] = geometry::Voxel(index, color);` in `cpp/open3d/pipelines/integration/UniformTSDFVolume.cpp`.
> 
> An easy fix would be to remove the OpenMP parallelization (fix 1.). Other possible fixes are (fix 2.) use `tbb:concurrent_unordered_map` which is thread-safe, but then you have to convert it back to a `std::unordered_map` (which I presume is O(n)?), or (fix 3.) use multiple `std::unordered_map`s, one for each thread, and then merge them after the parallelization (which would be O(n) average case). I have tested fix 1. and fix 2. for multiple voxel resolutions (16, 32, 64, 128, 256, 512, 768), and their timings are practically the same. [@ssheorey](https://github.com/ssheorey) do you have any other ideas?


Thanks for your reply, but the way I solved the problem was a bit crude.
Because I need to use this function `extract_voxel_grid` in both x86_64 and NVIDIA Jetson, 
the method I adopted is: 
after deleting [the conditional compilation program](https://github.com/isl-org/Open3D/blob/0f06a149c4fb9406fd3e432a5cb0c024f38e2f0e/cpp/open3d/pipelines/integration/UniformTSDFVolume.cpp#L263), I recompiled the python whl files of x86_64 and aarch64 of Open3D 0.18.0 in docker. It works well on both platforms.

When I tried Open3D 0.19.0 today, I still had the problem that extract_voxel_grid caused the program to terminate.
But I saw [this method](https://github.com/pearl-robot-lab/neugraspnet/issues/3#issuecomment-2746022017), So I added these at the beginning of the python program:
```
import os
os.environ['OMP_NUM_THREADS'] = '1'
```
The above problem did not occur again. Next I will try to use this method on aarch64.

**Comment by nicolaloi** [comment]
@hooyahliu yes, since the problem is caused by the parallelization, forcing the number of threads to 1 with the environmental variable is an easy fix from the user side. @ssheorey @benjaminum for an actual code fix, I proposed some solutions in [my previous comment](https://github.com/isl-org/Open3D/issues/6712#issuecomment-2119264223).

**Comment by ssheorey** [comment]
@nicolaloi thanks for looking into this. I think the best option for this would be to create per thread  `std::vector<geometry::Voxel>` initially in parallel with OpenMP, and then this can be inserted into the `std::unordered_map` representation one at a time.  (Same as your option 3, but we can use an `std::vector` since we don't need the `unordered_map` properties during construction). Please do submit a bugfix PR!
